### ISSUE TITLE
headerTemplate creates new outletContext on every change

### ISSUE THREAD

**Issue by arlowhite** [swimlane/ngx-datatable#785]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, post on Stackoverflow or Gitter
```

**Current behavior**

I use `headerTemplate` to put a material2 `MdInputContainer` in my column headers to provide a filter UI.
The `headerTemplate` regenerates far too often (on every change detection).  This resets the input text in my template as the user types. I worked around this by saving the input text externally, but even then there are serious issues.
In Firefox, when I select all text in the input, the entire Firefox browser crashes in Linux, and the tab crashes in Firefox Mac. I think elements are being destroyed while Firefox has a pointer to them in its text selection code.

Currently, the `headerTemplate` outlet context is created in the template like this:
```typescript
      <ng-template
        *ngIf="column.headerTemplate"
        [ngTemplateOutlet]="column.headerTemplate"
        [ngOutletContext]="{ 
          column: column, 
          sortDir: sortDir,
          sortFn: sortFn,
          allRowsSelected: allRowsSelected,
          selectFn: selectFn
        }">
      </ng-template>
```

I think this creates a new `ngOutletContext` object on every change detection, which causes the `headerTemplate` to be recreated.

**Expected behavior**

`headerTemplate` should only regenerate when properties it depends on change.

**Reproduction of the problem**

I don't have time to create a stand-alone demo at the moment.

**What is the motivation / use case for changing the behavior?**

Prevent `headerTemplate` from regenerating on every change detection.

**Please tell us about your environment:**

Node v6.9.1
npm 4.6.1

* **Table version:** 9.2.0

* **Angular version:** 4.1.3

```
 @angular/cli: 1.0.4
 node: 6.9.1
 os: linux x64
 @angular/animations: 4.1.3
 @angular/common: 4.1.3
 @angular/compiler: 4.1.3
 @angular/core: 4.1.3
 @angular/flex-layout: 2.0.0-beta.8
 @angular/forms: 4.1.3
 @angular/http: 4.1.3
 @angular/material: 2.0.0-beta.5-22a55aa
 @angular/platform-browser: 4.1.3
 @angular/platform-browser-dynamic: 4.1.3
 @angular/router: 4.1.3
 @angular/cli: 1.0.4
 @angular/compiler-cli: 4.1.3
```

* **Browser:** all

* **Language:** all

PR incoming.

**Comment by m-a-l-p** [comment]
Hi,

I'm currently using version 9.3.0 but still have this issue where my custom header gets destroyed every time onClick is triggered for sorting.

Here's how my codes look like:
`                         <ng-template ngx-datatable-header-template let-sort="sortFn">
                              <custom-table-header 
                                    [columnId]="currentColumn.dataIndex"
                                    [columnName]="currentColumn.columName"
                                    [currentColumn]="currentColumn"
                                    [tableObject]="tableObject"
                                    [tableId]="tableId"
                                    (click)="sort()">
                              </custom-table-header>
                        </ng-template>
`

I hope to get some advice from you.

Many thanks,
Toni

**Comment by rooterkyberian** [comment]
pinging @arlowhite , are you going to continue to work on this?

It seems I have hit the same thing when trying to implement column filters.
The issue seems to have appeared with 9.1.0 (so reverting to 9.0.0 helps) and exists still in 9.3.1.

Just quick and dirty example:
https://plnkr.co/edit/sjzQrt?p=preview

Try typing anything in "gender" input and double clicking to select whatever you typed in. Almost instantaneously header gets recreated, thus emptied and of course the selection itself is gone as well. Watch out - this can crash tabs in Firefox for some reason.

**Comment by amcdnl** [comment]
@arlowhite - did you fix this in a PR? I can't remember exactly.

**Comment by rooterkyberian** [comment]
@amcdnl I have merged arlowhite:fix-header-template-outlet-context with 9.3.1 and it did fix mentioned problem for me

**Comment by amcdnl** [comment]
@arlowhite - can we discuss your solution?

**Comment by arlowhite** [comment]
@amcdnl Rebased onto master (at tag 10.0.3) and created PR #903 

If you want to discuss/change my current implementation, I think we should discuss it in the PR so code comments can be used.

**Comment by amcdnl** [comment]
@arlowhite - I just did this change for cell component, you might get some ideas there to the implementation.

**Comment by arlowhite** [comment]
@amcdnl could you link the commit/PR and I'll look at it.

**Comment by wizarrc** [comment]
@arlowhite Here is the cell context commit. https://github.com/swimlane/ngx-datatable/commit/ce6c53370bfe0daccd6b253ca76e2e68f5ceb45e

**Comment by arlowhite** [comment]
My impression from reading angular.io is that ngDoCheck should be avoided unless it's really necessary.
https://angular.io/guide/lifecycle-hooks#docheck
> While the ngDoCheck() hook can detect when the hero's name has changed, it has a frightful cost. This hook is called with enormous frequency—after every change detection cycle no matter where the change occurred. It's called over twenty times in this example before the user can do anything.

**Comment by wizarrc** [comment]
@arlowhite `ngDoCheck()` is used in many common directives such as `NgStyle`, `NgSwitch`, `NgClass`, and `NgForOf` to check for changes. The use should be used when the alternative would cost more. In this case, using a literal object as the context will always create a new object and force the renderer to destroy and recreate a new view every change detection. I'm pretty sure the work inside `ngDoCheck()` is a much less than redraw everything every DOM event, such as scrolling, where performance is critical.

When the context is checked during the bindings where it calls [`checkBinding()`](https://github.com/angular/angular/blob/master/packages/core/src/view/util.ts#L81). Because the cell context is now static, it now does not see any changes, even when the properties inside the object have changed. `ngDoCheck` allows the library to check all the sub-properties, similar to how `@Input` properties work but all wrapped up into one property.

**Comment by arlowhite** [comment]
@amcdnl fixed this in 10.0.4
https://github.com/swimlane/ngx-datatable/commit/65c5afd37d10fa8da8d11302e05aaa6acc1a4ffd
